### PR TITLE
global sort: fix broken prop with onefile writer

### LINKED ISSUE #49590: global sort read wrong offset when read stat file of onefile writer
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
run 1T import into with global sort

### 2. What did you expect to see? (Required)
import success.
### 3. What did you see instead (Required)
```
[2023/12/18 11:50:21.137 +00:00] [Info] [reader.go:42] [arguments] [data-file-count=34] [stat-file-count=34] [start-key="dIAAAAAAAABwX2mAAAAAAAAACgQAAAAH/eTrQwExMDYwNTY2N/8xRk14N2ZPTf80QlhlZlNiNP81Skh4bmYzZf9iYmhaUkJQQ/9xAAAAAAAAAPg="] [end-key="dIAAAAAAAABwX2mAAAAAAAAACgQAAAAH/eTraQExODI5MjQ0MP83SG9yY3RQU/9hNFptRjhSTP84RUF3dE1sUP9Ia3UweFN2bf8wAAAAAAAAAPg="]

[2023/12/18 11:50:21.866 +00:00] [Info] [engine.go:197] ["found hotspot file in getFilesReadConcurrency"] [filename=240002/360025/36e833b5-093a-412f-b8d8-12405020fed2_stat/one-file] [startOffset=643824000] [endOffset=0] ["expected concurrency"=4398046510950]
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #49590 

Problem Summary:

### What changed and how does it work?
When exceed writer memlimit, create new prop with kv.offset.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/onefile_writer.go [hunk 1]
```diff
@@ -103,6 +105,8 @@ func (w *OneFileWriter) WriteRow(ctx context.Context, idxKey, idxVal []byte) err
 			return err
 		}
 		w.rc.reset()
+		// the new prop should have the seem offset with kvStore.
```
Comment (ywqzzy) [inline br/pkg/lightning/backend/external/onefile_writer.go:108]: The bug fix is here

Comment (Benjamin2037) [reply br/pkg/lightning/backend/external/onefile_writer.go:108]:
```suggestion
		// the new prop should have the same offset with kvStore.
```

File: br/pkg/lightning/backend/external/sort_test.go [hunk 2]
```diff
@@ -148,7 +148,7 @@ func TestGlobalSortLocalWithMerge(t *testing.T) {
 		startKey = BytesMin(startKey, s.Min.Clone())
 		endKey = BytesMax(endKey, s.Max.Clone().Next())
 	}
-
+	mergeMemSize := (rand.Intn(10) + 1) * 100
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/sort_test.go:151]: what this “mergeMemSize := (rand.Intn(10) + 1) * 100” means，why here use a random func?

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: br/pkg/lightning/backend/external/sort_test.go (hunk 2 above)
Comment (ywqzzy) [reply br/pkg/lightning/backend/external/sort_test.go:151]: Make the test run more cases to check different memSize/blockSize of writer can be used for writer.
When memSize is small enough, we will reach line 109 in onefile writer, hopefully the bug will reproduce.

File: br/pkg/lightning/backend/external/bench_test.go [hunk 3]
```diff
@@ -761,6 +763,7 @@ var (
 	concurrency  = flag.Int("concurrency", 100, "concurrency")
 	memoryLimit  = flag.Int("memory-limit", 64*units.MiB, "memory limit")
 	skipCreate   = flag.Bool("skip-create", false, "skip create files")
+	fileName     = flag.String("file-name", "test", "test")
```
Comment (ywqzzy) [inline br/pkg/lightning/backend/external/bench_test.go:766]: oops, I will rename it

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**